### PR TITLE
Keep track of whether the user has a real password set

### PR DESCRIPTION
**Pull request by bl-nero**
This change addresses most of the aspects of RFD 159, sections [Recognizing Users Who Configured Their Passwords](https://github.com/gravitational/teleport/blob/master/rfd/0159-password-status.md#recognizing-users-who-configured-their-passwords) and [Changes to the password reset flow](https://github.com/gravitational/teleport/blob/master/rfd/0159-password-status.md#changes-to-the-password-reset-flow).

Updating the flag on successful login hasn't yet been implemented, I'm working on it in a separate PR.

### CODE REVIEW

**Comment by bl-nero** [comment]
Following cases were tested:
1. Creating a user using `tctl` and Web.
2. Changing password.
3. Resetting password using `tctl` and web.
4. Creating a bot user using `tctl`.
5. Password recovery.

**Comment by bl-nero** [comment]
Dear reviewers, I need an advice about what to do with the [unit test failure](https://github.com/gravitational/teleport/actions/runs/8346953796/job/22845454710?pr=39576): should I modify the test, assuming that the new default password state should indeed be default, or should it be a signal that I broke some important assumption?

**Comment by espadolini** [comment]
> Dear reviewers, I need an advice about what to do with the [unit test failure](https://github.com/gravitational/teleport/actions/runs/8346953796/job/22845454710?pr=39576): should I modify the test, assuming that the new default password state should indeed be default, or should it be a signal that I broke some important assumption?

That looks exactly like the sort of thing that the `status` field is for:
> If a resource has properties that are required to be modified dynamically by Teleport, a separate `status` field should be added to the resource to contain them. These fields will be ignored by IaC tools during their reconcilliation.

It's not quite IaC tools, but it seems like `TestCreateResources/user` should be ignoring `status` in that check.

**Comment by bl-nero** [comment]
@espadolini Thanks for confirming. Fixing it.

**Comment by bl-nero** [comment]
I'm converting this one to draft, because there are still some scenarios in which we risk breaking the data model consistency, like this:

1. Process _A_ intends to delete a password. It updates the state to UNSPECIFIED, generating revision _a_.
2. Process _B_ intends to update a password. It updates the state to UNSPECIFIED, generating revision _b_.
3. Process _B_ updates the password hash.
4. Process _A_ removes the password hash.
5. Process _B_ sets the password state to SET, asserting on revision _b_, and succeeds.
6. Process _A_ attempts to set the password state to UNSET, asserting on revision _a_, and fails.

In this scenario, the password is left unset, but the password state is not UNSPECIFIED, it's SET.

I think we won't escape using atomic operations.

**Comment by zmb3** [comment]
Why set the password state to `UNSPECIFIED` first?

I would just set it to `SET` after the password reset/login completes successfully. That should be good enough. Still some races, but there shouldn't be any false positives, only false negatives, which is the status quo.

**Comment by bl-nero** [comment]
OK, I'll just simplify this and ignore the race conditions.

**Comment by bl-nero** [comment]
Modified according to @zmb3's suggestions, ready for review.

File: api/types/user.go [hunk 1]
```diff
@@ -139,6 +139,19 @@ type User interface {
...
+	// perform any password-related activity since then. See RFD 0159 for details.
+	// Do NOT use this value for authentication purposes!
+	GetPasswordState() PasswordState
+	// SetPasswordState updates the information about user's password. Note that
+	// this is a "best effort" property, in that it can be UNSPECIFIED for users
+	// who were created before this property was introduced and didn't perform any
+	// password-related activity since then. See RFD 0159 for details.
+	SetPasswordState(PasswordState)
+	// DeepCopy creates a clone of this User value.
+	DeepCopy() User
```
Comment (espadolini) [inline api/types/user.go:154]: I think we use `Clone()` for a lot of types in `api/types` already.

File: lib/auth/usertoken.go [hunk 2]
```diff
@@ -148,7 +148,7 @@ func (a *Server) CreateResetPasswordToken(ctx context.Context, req CreateUserTok
 		return nil, trace.BadParameter("invalid reset password token request type")
 	}
 
-	_, err = a.ResetPassword(ctx, req.Name)
+	err = a.ResetPassword(ctx, req.Name)
 	if err != nil {
```
Comment (espadolini) [inline lib/auth/usertoken.go:152]: Here and in the other places that you've touched, might as well clean up the code a bit.

```suggestion
	if err := a.ResetPassword(ctx, req.Name); err != nil {
```

File: lib/services/authentication.go [hunk 3]
```diff
@@ -31,13 +31,22 @@ import (
...
 func ValidateLocalAuthSecrets(l *types.LocalAuthSecrets) error {
+	var hash []byte
 	if len(l.PasswordHash) > 0 {
-		if _, err := bcrypt.Cost(l.PasswordHash); err != nil {
-			return trace.BadParameter("invalid password hash")
-		}
+		hash = l.PasswordHash
+	} else {
+		// Prevent timing attacks by using a fake hash.
+		hash = fakePasswordHash
```
Comment (espadolini) [inline lib/services/authentication.go:44]: Why is this necessary now? Was this always necessary and we just never bothered?

File: lib/services/identity.go [hunk 4]
```diff
@@ -111,9 +108,15 @@ type Identity interface {
...
 	GetUsedTOTPToken(user string) (string, error)
 
-	// UpsertPassword upserts new password and OTP token
+	// UpsertPassword upserts a new password. It also sets the user's
+	// `PasswordState` status flag accordingly. Returns an error if the user
+	// doesn't exist.
 	UpsertPassword(user string, password []byte) error
 
+	// DeletePassword deletes user's password and sets the `PasswordState` status
+	// flag accordingly. Does NOT return an error if the password didn't exist.
```
Comment (espadolini) [inline lib/services/identity.go:117]: Why not bubble up the `NotFound` and only ignore it in `ResetPassword`?

**Review by espadolini — APPROVED** [review]

**Comment by bl-nero** [comment]
@zmb3 This PR needs ignoring the flaky test detector due to another timeout. Note that I tried to do the same thing as @codingllama did in https://github.com/gravitational/teleport.e/pull/3551/files, but somehow, lowering the bcrypt cost here had no effect on the test duration.

**Comment by bl-nero** [comment]
Apologies, I accidentally pushed a commit that was supposed to end on another branch, so I force-pushed the branch to its original state.

File: lib/services/local/users.go [hunk 5]
```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
...
 	}
 
-	err = s.UpsertPasswordHash(user, hash)
-	if err != nil {
+	if err := s.upsertPasswordHash(user, hash); err != nil {
 		return trace.Wrap(err)
 	}
 
+	_, err = s.UpdateAndSwapUser(
+		context.TODO(),
```
Comment (zmb3) [inline lib/services/local/users.go:815]:
```suggestion
		context.Background(),
```

I'd reserve TODO for placeholders that you intend to go back and address, and `Background` for places where you explicitly want a new context.

**Review by zmb3 — APPROVED** [review]

File: lib/auth/usertoken.go [hunk 6]
```diff
@@ -148,8 +148,13 @@ func (a *Server) CreateResetPasswordToken(ctx context.Context, req CreateUserTok
 		return nil, trace.BadParameter("invalid reset password token request type")
 	}
 
-	_, err = a.ResetPassword(ctx, req.Name)
-	if err != nil {
+	// Since ResetPassword ignores doesn't report errors for non-existent users,
+	// first check if the user exists.
```
Comment (codingllama) [inline lib/auth/usertoken.go:152]: There are a few extra words in there, rephrase?

```suggestion
	// ResetPassword doesn't error for non-existent users, so first check if the
	// user exists.
```

File: lib/auth/password_test.go [hunk 7]
```diff
@@ -845,6 +846,34 @@ func TestChangeUserAuthenticationWithErrors(t *testing.T) {
...
+
+	// Reset password.
+	ctx := context.Background()
+	err = s.a.ResetPassword(ctx, "dave")
+	require.NoError(t, err)
+
+	// Make sure that the password has been reset.
+	u, err := s.a.Identity.GetUser(ctx, "dave", true /* withSecrets */)
+	require.NoError(t, err)
+	assert.Nil(t, u.GetLocalAuth())
```
Comment (codingllama) [inline lib/auth/password_test.go:868]:
```suggestion
	assert.Nil(t, u.GetLocalAuth(), "user LocalAuth not nil")
```

File: lib/services/authentication.go (hunk 3 above)
Comment (codingllama) [reply lib/services/authentication.go:44]: What are we trying to accomplish here? Its' not clear to me what the intent is.

File: lib/services/authentication_test.go [hunk 8]
```diff
@@ -63,3 +64,20 @@ func TestValidateLocalAuthSecrets_deviceTypes(t *testing.T) {
...
+	assert.NoError(t, services.ValidateLocalAuthSecrets(&types.LocalAuthSecrets{
+		// bcrypt hash of "foobar"
+		PasswordHash: []byte("$2y$10$d3BZ9tUDA5vD1hUL8iSfC.ADGj.WS4VRTLVtEWkZrD76pRZFJZ5f2"),
+	}))
+
+	err := services.ValidateLocalAuthSecrets(&types.LocalAuthSecrets{
+		PasswordHash: []byte("$hashimpo$tor"),
+	})
+	assert.Error(t, err)
+	assert.True(t, trace.IsBadParameter(err))
```
Comment (codingllama) [inline lib/services/authentication_test.go:82]:
```suggestion
	assert.True(t, trace.IsBadParameter(err), "ValidateLocalAuthSecrets returned err=%v (%T), want BadParameter", err, trace.Unwrap(err))
```

File: lib/services/local/users.go [hunk 9]
```diff
@@ -287,6 +287,13 @@ func (s *IdentityService) CreateUser(ctx context.Context, user types.User) (type
 		return nil, trace.AlreadyExists("user %q already registered", user.GetName())
 	}
 
+	auth := user.GetLocalAuth()
+	if auth != nil && len(auth.PasswordHash) > 0 {
```
Comment (codingllama) [inline lib/services/local/users.go:291]: When do we expect the password to be set here, vs unset? Is it unset in prod but set for testing? Please add a brief comment clarifying.

File: lib/services/local/users.go [hunk 10]
```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
...
+		context.TODO(),
+		user,
+		false, /*withSecrets*/
+		func(u types.User) (bool, error) {
+			u.SetPasswordState(types.PasswordState_PASSWORD_STATE_SET)
+			return true, nil
+		})
+	if err != nil {
+		// Don't let the password state flag change fail the entire operation.
+		s.log.WithError(err).Info("Failed to set password state")
```
Comment (codingllama) [inline lib/services/local/users.go:824]: Bump to warning, log user name?

```suggestion
		s.log.
			WithError(err).
			WithField("User", user).
			Warn("Failed to set user password state")
```

File: lib/services/local/users.go [hunk 11]
```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
...
+			context.TODO(),
+			user,
+			false, /*withSecrets*/
+			func(u types.User) (bool, error) {
+				u.SetPasswordState(types.PasswordState_PASSWORD_STATE_UNSET)
+				return true, nil
+			})
+		if err != nil {
+			// Don't let the password state flag change fail the entire operation.
+			s.log.WithError(err).Info("Failed to set password state")
```
Comment (codingllama) [inline lib/services/local/users.go:850]: Same as above.

File: lib/services/local/users.go [hunk 12]
```diff
@@ -800,14 +807,53 @@ func (s *IdentityService) UpsertPassword(user string, password []byte) error {
...
+		}
+	}
+
+	return trace.Wrap(delErr)
```
Comment (codingllama) [inline lib/services/local/users.go:854]: Suggestion: return early to reduce nesting:

```suggestion
	err := s.Delete(ctx, backend.Key(webPrefix, usersPrefix, user, pwdPrefix)
	if err != nil && !trace.IsNotFound(err) {
		return trace.Wrap(err)
	}

	if _, err := s.UpdateAndSwapUser(
		context.TODO(),
		user,
		false, /*withSecrets*/
		func(u types.User) (bool, error) {
			u.SetPasswordState(types.PasswordState_PASSWORD_STATE_UNSET)
			return true, nil
		},
	); err != nil {
		// Don't let the password state flag change fail the entire operation.
		s.log.WithError(err).Info("Failed to set password state")
	}
	
	return nil
```

File: lib/services/local/users_test.go [hunk 13]
```diff
@@ -1177,6 +1224,53 @@ func TestIdentityService_ListUsers(t *testing.T) {
...
+	assert.Equal(t, types.PasswordState_PASSWORD_STATE_UNSET, user.GetPasswordState(), "password state is not UNSET")
+
+	// Set the password.
+	err = identity.UpsertPassword(user.GetName(), []byte("gimme12chars"))
+	require.NoError(t, err, "failed to set password")
+
+	// Validate that the password is set.
+	user, err = identity.GetUser(ctx, "capybara", true /* withSecrets */)
+	require.NoError(t, err, "failed to get the user")
+	bcrypt.CompareHashAndPassword(user.GetLocalAuth().PasswordHash, []byte("gimme12chars"))
```
Comment (codingllama) [inline lib/services/local/users_test.go:1250]:
```suggestion
	err = bcrypt.CompareHashAndPassword(user.GetLocalAuth().PasswordHash, []byte("gimme12chars"))
```

File: lib/services/local/users_test.go [hunk 14]
```diff
@@ -60,6 +60,53 @@ func newIdentityService(t *testing.T, clock clockwork.Clock) *local.IdentityServ
...
+			}
+			_, err = identity.CreateUser(ctx, user)
+			require.NoError(t, err, "failed to create the user")
+
+			user, err = identity.GetUser(ctx, tt.username, true /* withSecrets */)
+			require.NoError(t, err, "failed to get the user")
+			assert.Equal(t, []string{tt.role}, user.GetRoles(), "roles are not set")
+			assert.Equal(t, tt.passwordState, user.GetPasswordState(), "password state is not set")
+			if tt.password != "" {
+				bcrypt.CompareHashAndPassword(user.GetLocalAuth().PasswordHash, []byte(tt.password))
```
Comment (codingllama) [inline lib/services/local/users_test.go:102]:
```suggestion
				err = bcrypt.CompareHashAndPassword(user.GetLocalAuth().PasswordHash, []byte(tt.password))
```

Or inline it right into the assert, which I find a bit better in cases like this (less risk to miss the return).

File: lib/services/authentication.go (hunk 3 above)
Comment (bl-nero) [reply lib/services/authentication.go:44]: @codingllama I think it was a miscommunication on my side, I was under impression that either you or perhaps Mike Jensen claimed that this branch should also be protected from timing attacks. I don't really think myself that it's necessary, so I'm removing this part.

File: lib/services/local/users.go (hunk 12 above)
Comment (codingllama) [reply lib/services/local/users.go]: I just flipped the position of some of the ifs, but the code should still be equivalent. If it's not equivalent than I made an editing mistake, but nevertheless the style can be changed. Please take a another look, but if you'd rather not apply it it's also fine, it's only a suggestion.

Comment (bl-nero) [reply lib/services/local/users.go]: Oh, of course, it can be still done with an early return, I only need to check for this error twice. :) Got it.

**Review by codingllama — APPROVED** [review]

**Comment by bl-nero** [comment]
Merged with master, retested.

**Comment by bl-nero** [comment]
I had to push a fix to a timing issue inside Teleterm tests.
